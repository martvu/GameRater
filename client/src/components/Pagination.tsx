import { Button } from './ui/button';

interface PaginationProps {
  currentPage: number;
  setCurrentPage: (currentPage: number) => void;
  pages: number;
  storageKey: 'gameListPage' | 'reviewPage';
}

/**
 * Pagination component
 * @param {number} currentPage - The current page
 * @param {function} setCurrentPage - Callback function when page is changed
 * @param {number} pages - The total number of pages
 */
export default function Pagination({
  currentPage,
  setCurrentPage,
  pages,
  storageKey,
}: PaginationProps) {
  interface PageButtonProps {
    pageNumber: number;
  }

  function handlePageChange(pageNumber: number) {
    setCurrentPage(pageNumber);
    // Save page number to session storage, only for GameListPage
    if (storageKey === 'gameListPage') {
      sessionStorage.setItem(storageKey, pageNumber.toString());
    }
  }

  const PageButton = ({ pageNumber }: PageButtonProps) => {
    const isActive = currentPage === pageNumber;
    return (
      <Button
        variant={isActive ? 'outline' : 'ghost'}
        className="border-primary"
        size="page"
        onClick={() => handlePageChange(pageNumber)}
        aria-label={`go to page ${pageNumber}`}
      >
        {pageNumber}
      </Button>
    );
  };

  if (pages > 1)
    return (
      <nav aria-label="pagination" className="flex flex-wrap justify-center">
        <Button
          size="page"
          disabled={currentPage === 1}
          onClick={() => handlePageChange(currentPage - 1)}
          variant="ghost"
          aria-label="go to previous page"
        >
          Â«
        </Button>
        {pages <= 7 ? (
          Array<number>(pages)
            .fill(1)
            .map((_, index) => {
              return <PageButton pageNumber={index + 1} key={index} />;
            })
        ) : currentPage <= 4 ? (
          <>
            <PageButton pageNumber={1} />
            <PageButton pageNumber={2} />
            <PageButton pageNumber={3} />
            <PageButton pageNumber={4} />
            <PageButton pageNumber={5} />
            <Button
              size="page"
              variant="ghost"
              disabled
              aria-label="more pages"
            >
              ...
            </Button>
            <PageButton pageNumber={pages} />
          </>
        ) : currentPage >= pages - 3 ? (
          <>
            <PageButton pageNumber={1} />
            <Button
              size="page"
              variant="ghost"
              disabled
              aria-label="more pages"
            >
              ...
            </Button>
            <PageButton pageNumber={pages - 4} />
            <PageButton pageNumber={pages - 3} />
            <PageButton pageNumber={pages - 2} />
            <PageButton pageNumber={pages - 1} />
            <PageButton pageNumber={pages} />
          </>
        ) : (
          <>
            <PageButton pageNumber={1} />
            <Button
              size="page"
              variant="ghost"
              disabled
              aria-label="more pages"
            >
              ...
            </Button>
            <PageButton pageNumber={currentPage - 1} />
            <PageButton pageNumber={currentPage} />
            <PageButton pageNumber={currentPage + 1} />
            <Button
              size="page"
              variant="ghost"
              disabled
              aria-label="more pages"
            >
              ...
            </Button>
            <PageButton pageNumber={pages} />
          </>
        )}
        <Button
          size="page"
          disabled={currentPage === pages}
          onClick={() => handlePageChange(currentPage + 1)}
          variant="ghost"
          aria-label="go to next page"
        >
          Â»
        </Button>
      </nav>
    );
}
