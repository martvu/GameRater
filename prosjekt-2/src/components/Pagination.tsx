import { ChevronRight, ChevronLeft } from 'lucide-react';
import { Button } from './ui/button';

interface PaginationProps {
  currentPage: number;
  setCurrentPage: (currentPage: number) => void;
  gamesPerPage: number;
  gameData: unknown[];
}
export default function Pagination({
  currentPage,
  setCurrentPage,
  gamesPerPage,
  gameData,
}: PaginationProps) {
  const totalPages = Math.ceil(gameData.length / gamesPerPage);

  const nextPage = () => {
    if (currentPage < totalPages) {
      setCurrentPage(currentPage + 1);
    }
  };

  const prevPage = () => {
    if (currentPage > 1) {
      setCurrentPage(currentPage - 1);
    }
  };

  return (
    <nav role="navigation" aria-label="Pagination Navigation">
      <ul className="flex list-none items-center justify-center text-sm md:gap-1">
        <li>
          <Button
            variant="ghost"
            onClick={prevPage}
            aria-label="Go to Previous Page"
            disabled={currentPage === 1}
          >
            <span className="order-2">Prev</span>
            <ChevronLeft className="h-4 w-4" />
          </Button>
        </li>
        
        <li>
          <span className="mx-2">{`${currentPage} / ${totalPages}`}</span>
        </li>

        <li>
          <Button
            variant="ghost"
            onClick={nextPage}
            aria-label="Go to Next Page"
            disabled={currentPage === totalPages}
          >
            <span>Next</span>
            <ChevronRight className="h-4 w-4" />
          </Button>
        </li>
      </ul>
    </nav>
  );
}
