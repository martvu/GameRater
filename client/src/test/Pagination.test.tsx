import Pagination from '@/components/Pagination';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, vi } from 'vitest';
import userEvent from '@testing-library/user-event';

describe('Pagination', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <Pagination currentPage={1} setCurrentPage={() => {}} pages={5} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly', () => {
    render(<Pagination currentPage={1} setCurrentPage={() => {}} pages={5} />);
    const pageButtons = screen.getAllByRole('button');
    // 5 pages + 2 for previous and next buttons
    expect(pageButtons).toHaveLength(7);
  });

  it('calls setCurrentPage with the correct page number', () => {
    const setCurrentPageMock = vi.fn();
    render(
      <Pagination
        currentPage={1}
        setCurrentPage={setCurrentPageMock}
        pages={5}
      />
    );
    const secondPageButton = screen.getByLabelText('go to page 2');
    fireEvent.click(secondPageButton);
    expect(setCurrentPageMock).toHaveBeenCalledWith(2);
  });

  it('increments page number on next button click', async () => {
    const setCurrentPage = vi.fn();
    render(
      <Pagination currentPage={1} setCurrentPage={setCurrentPage} pages={5} />
    );

    const nextButton = screen.getByLabelText('go to next page');
    await userEvent.click(nextButton);

    expect(setCurrentPage).toHaveBeenCalledWith(2);
  });

  it('decrements page number on previous button click', async () => {
    const setCurrentPage = vi.fn();
    render(
      <Pagination currentPage={2} setCurrentPage={setCurrentPage} pages={5} />
    );

    const prevButton = screen.getByLabelText('go to previous page');
    await userEvent.click(prevButton);

    expect(setCurrentPage).toHaveBeenCalledWith(1);
  });

  it('disables the previous button on the first page', () => {
    render(<Pagination currentPage={1} setCurrentPage={() => {}} pages={5} />);
    const previousButton = screen.getByLabelText('go to previous page');
    expect(previousButton).toBeDisabled();
  });

  it('disables the next button on the last page', () => {
    render(<Pagination currentPage={5} setCurrentPage={() => {}} pages={5} />);
    const nextButton = screen.getByLabelText('go to next page');
    expect(nextButton).toBeDisabled();
  });
});
