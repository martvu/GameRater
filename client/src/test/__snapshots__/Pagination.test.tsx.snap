// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="pagination"
  >
    <div
      class="flex flex-wrap justify-center"
    >
      <button
        aria-label="go to previous page"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10"
        disabled=""
      >
        «
      </button>
      <button
        aria-label="go to page 1"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border bg-background hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10 border-primary"
      >
        1
      </button>
      <button
        aria-label="go to page 2"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10 border-primary"
      >
        2
      </button>
      <button
        aria-label="go to page 3"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10 border-primary"
      >
        3
      </button>
      <button
        aria-label="go to page 4"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10 border-primary"
      >
        4
      </button>
      <button
        aria-label="go to page 5"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10 border-primary"
      >
        5
      </button>
      <button
        aria-label="go to next page"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground w-9 h-10 sm:h-12 sm:w-10"
      >
        »
      </button>
    </div>
  </nav>
</DocumentFragment>
`;
